File: services/orchestrator/internal/entity/proof.go
// Package entity holds Solva's pure domain objects. It imports nothing
// outside the standard library so the inner layer stays free of frameworks,
// drivers, and transport concerns.
package entity

import "time"

// Proof is one published solvency proof for a tenant. Money fields are
// decimal strings of integer minor units, never floats, so values round-trip
// exactly between the prover, the chain, and Postgres.
type Proof struct {
	// ID is the orchestrator's internal proof identifier.
	ID string
	// TenantID owns this proof.
	TenantID string
	// ChainProofID is the monotonic id returned by the Soroban registry.
	ChainProofID uint64
	// RootHash is the Poseidon2 sum-tree root, hex. Computed by the prover,
	// not by Go. The orchestrator only carries it.
	RootHash string
	// ReservesTotal (R) and LiabilitiesTotal (L) are decimal strings.
	ReservesTotal    string
	LiabilitiesTotal string
	// Blob is the serialized UltraHonk proof.
	Blob []byte
	// PublishedAt is the on-chain timestamp of the proof.
	PublishedAt time.Time
}

// PublicInputs are the values exposed on-chain for a proof. Everything else in
// the witness stays private to the prover.
type PublicInputs struct {
	// ReservesTotal (R), LiabilitiesTotal (L), and PrevReserves (R_prev) are
	// decimal strings of integer minor units.
	ReservesTotal    string
	LiabilitiesTotal string
	PrevReserves     string
	// RootHash is the Poseidon2 sum-tree root, hex.
	RootHash string
}

File: packages/shared-types/src/proof.ts
// Types that mirror proto/prover.proto. All monetary and field values are
// decimal or hex strings, never numbers, so we never lose precision to floats.

/** A single customer liability committed into the Merkle Sum Tree. */
export interface Liability {
  /** Poseidon2 hash of the external customer reference, hex. */
  customerIdHash: string;
  /** Balance in integer minor units, decimal string. */
  balance: string;
}

/** A single reserve figure attested by a bank or on-chain source. */
export interface Reserve {
  sourceId: string;
  /** Balance in integer minor units, decimal string. */
  balance: string;
}

/**
 * Values exposed on-chain by a published proof. Everything else stays private.
 * Maps to proto PublicInputs.
 */
export interface PublicInputs {
  /** R, total reserves, decimal string. */
  reservesTotal: string;
  /** L, total liabilities, decimal string. */
  liabilitiesTotal: string;
  /** root_h, Poseidon2 sum-tree root, hex. */
  rootHash: string;
  /** R_prev, previous cycle reserve total, used for the fraud bound. */
  prevReserves: string;
}

/**
 * A published proof as the SDK and web app see it. The proof bytes are the
 * serialized UltraHonk proof, base64 encoded for transport over REST.
 */
export interface Proof {
  /** Monotonic on-chain proof ID. */
  id: string;
  /** Base64 of the serialized UltraHonk proof. */
  proof: string;
  publicInputs: PublicInputs;
  /** Unix epoch seconds when the proof was published on-chain. */
  publishedAt: number;
}
